### PR TITLE
ILP for pathfinding/routing

### PR DESCRIPTION
**Pull request by makslevental**
This PR implements routing/path-finding using ILP (using Gurobi/or-tools) through the python pass registration mechanism landed in https://github.com/Xilinx/mlir-aie/pull/710. 

Background literature on the (static) Traffic Assignment problem can be found [here](https://sboyles.github.io/book.pdf#page=85) and [here](https://link.springer.com/chapter/10.1007/978-3-030-66501-2_41). Brief summary of the model follows; define:

* $(i,j) \in A$ : edges/wires between switchboxes;
* $\mathbf{h}$ : all possible flows, indexed by possible paths $\pi \in \Pi(r, s)$ from path-endpoint (source) $r$  to path-endpoint (target) $s$;
* $d^{rs}$ : number of flows we need to route from $r$  to $s$;
* $x_{ij}$ : amount of flow across an edge/link/wire;
* $t_{ij}(x_{ij})$ : the link performance/penalty functions.

Then the optimization problems is

```math
\begin{align}
\min_{\mathbf{x}, \mathbf{h}} & & \sum_{\mathbf{h}} t_{ij}(x_{ij})\cdot x_{ij} && (1) \\
\text{s.t.} & & x_{ij} = \sum_{\pi \in \Pi} \delta_{ij}^\pi h^\pi && (2) \\
&& d^{rs} = \sum_{\pi \in \Pi(r ,s)} h^\pi && (2) \\
&& x_{ij} \in \{0, \dots, c_{ij} \} && (3) 
\end{align}
```

where 

1. the objective to minimze, is the Total System Sime (TST), i.e., can be read as  "minimize distance $\times$ time for all flows";
2. flow constraints - flow across each edge is the sum of contributions from all flows with paths passing through that edge;
3. all requested flows must be routed;
4. each edge must not exceed its capacity.

The above model is (roughly) encoded in two modeling languages (Gurobi and or-tools), in [utils.py](https://github.com/Xilinx/mlir-aie/pull/761/files#diff-6a7ef7dc1545fbe9e571a1315dd52d6da9327e737bb07a74826fb13ff0ed1ed6R143), and tested in CI (only or-tools). Furthermore, in the or-tools formulation, another objective is implemented (minimum number of global edges) in order to compare/constrast with the existing `AIEPathfinder`.

I am still iterating on the pieces/parts/places but the core (the modeling) is done (I believe) and it would be good to get feedback on the design ASAP.

### CODE REVIEW

File: python/requirements.txt
Comment (jsetoain) [inline python/requirements.txt:4]: ?

File: test/python/python_passes.py
Comment (jsetoain) [inline test/python/python_passes.py:145]: Are these (and the ones below) supposed to be `# CHECK:` directives?

**Review by jsetoain — COMMENTED** [review]
Great work, thanks! I'll let @erwei-xilinx or @fifield have the last word, but it looks good to me.

**Review by erwei-xilinx — APPROVED** [review]
This looks really cool. Thanks.

File: python/requirements.txt
Comment (makslevental) [reply python/requirements.txt:4]: no clue why this got lit up - maybe space at EOL?

File: test/python/python_passes.py
Comment (makslevental) [reply test/python/python_passes.py:145]: yup (whoops)

File: python/AIEMLIRModule.cpp
Comment (makslevental) [reply python/AIEMLIRModule.cpp:32]: These were clang-tidy suggestions that probably recommends everything that's not mutated in the body be `const`. Why shouldn't it be `const`?

File: .github/workflows/buildAndTest.yml
Comment (makslevental) [reply .github/workflows/buildAndTest.yml:112]: Not 3.11 but I would like at least 3.9 for [Type Hinting Generics in Standard Collections](https://docs.python.org/3/whatsnew/3.9.html#type-hinting-generics-in-standard-collections) and ideally 3.10 for [writing union types as `X | Y`](https://peps.python.org/pep-0604/). To be honest, I do not see the value of being behind current main release of python (soon to be 3.12). We're talking about MLIR here, which is already bleeding edge technology and has basically no notion of semver.

Comment (keryell) [reply .github/workflows/buildAndTest.yml:112]: I agree for modern Python and C++23 at the same time. ;-)

File: python/util.py
Comment (keryell) [inline python/util.py:242]: I was wondering how it was possible to have a solver used in the neighborhood without notice...
https://peps.python.org/pep-0008/#imports says:
> Imports are always put at the top of the file, just after any module comments and docstrings, and before module globals and constants.

Any serious reason to deviate?
And actually it looks like there are a lot of other `import` all over the place.

Comment (makslevental) [reply python/util.py:242]: Importing within function scope is very common exactly for this purpose - you do not want to trigger an `ImportError` for this file (generically called `util.py`) just because Gurobi isn't installed if the user never tries to run any of the ILP related functions.

File: .github/workflows/buildAndTest.yml
Comment (jsetoain) [reply .github/workflows/buildAndTest.yml:112]: I'm personally always running the latest version, but in a public project we need to be a bit more reasonable. As a rule of thumb, I'd stay at most at whatever version ships with the oldest fully supported Ubuntu LTS (3.8.2 at the moment). The more complicated you make it for users to run your software, the fewer of them will stick with it. And if you expect it to be deployed within any company with any sort of software policy and IT department, you can forget about it.

File: python/AIEMLIRModule.cpp
Comment (jsetoain) [reply python/AIEMLIRModule.cpp:32]: It's not that it should, it's just that it doesn't make a difference 🤷 I don't care either way, I was just curious.

File: .github/workflows/buildAndTest.yml
Comment (keryell) [reply .github/workflows/buildAndTest.yml:112]: But is it our goal to promote implicitly technical debt by supporting almost 6 year old LTS?
Just advise to use a VM to run in an antique shop or to move to a company interested by software? :-)

File: python/util.py
Comment (keryell) [reply python/util.py:245]: That sounds like a compelling use case.
Why not adding a comment about this?

File: .github/workflows/buildAndTest.yml
Comment (makslevental) [reply .github/workflows/buildAndTest.yml:112]: > but in a public project we need to be a bit more reasonable.

This public project depends on MLIR which has no semver (as I mentioned). As I also mentioned, I being basically the core maintainer of upstream Python bindings, am always pushing to bump the supported Python version because Python has no breaking changes between 3.8 and 3.12 (so upgrading is "free"). So in summary: IMHO, it doesn't make sense to be conservative downstream of an unstable upstream.

Comment (jsetoain) [reply .github/workflows/buildAndTest.yml:112]: > But is it our goal to promote implicitly technical debt by supporting almost 6 year old LTS? Just advise to use a VM to run in an antique shop or to move to a company interested by software? :-)

You mean like ours? Because we do run 20.04 LTS in all our workstations 😂 

> This public project depends on MLIR which has no semver (as I mentioned). As I also mentioned, I being basically the core maintainer of upstream Python bindings, [am always pushing to bump the supported Python version](https://github.com/llvm/llvm-project/pull/68009) because Python has no breaking changes between 3.8 and 3.12 (so upgrading is "free"). So I can reduce the version here but in a matter of months upstream (me) will bump the required version anyway.
> 
> So in summary: IMHO, it doesn't make sense to be conservative downstream of an unstable upstream.

I may be mistaken here, but I thought LLVM required Python >=3.6, I'd expect everything to compile and run with 3.6 (tangentially, I bet that was 18.4 LTS included version, which went out of full support earlier this year). Ultimately, I'm just making a suggestion here. It's good practice to be (reasonably) conservative with dependency versions when your software is intended for wider adoption. And just to be absolutely clear, that's not my preference (I update the dependencies of all my personal projects every time I do anything with them, and I run Debian Sid in my PC), that's just a fact of life 🤷

Comment (makslevental) [reply .github/workflows/buildAndTest.yml:112]: I don't know about LLVM but MLIR currently supports >=3.8 (but not for long).

Comment (jsetoain) [reply .github/workflows/buildAndTest.yml:112]: Being part of LLVM, I expected requirements to be aligned. TIL.

Comment (keryell) [reply .github/workflows/buildAndTest.yml:112]: > You mean like ours? Because we do run 20.04 LTS in all our workstations 😂

Oh, I thought you were talking about external users. :-)
Just ask IT to upgrade to 22.04 and/or to install whatever Python version?

File: python/PythonPathFinderPass.cpp
Comment (stephenneuendorffer) [inline python/PythonPathFinderPass.cpp:35]: "Pathfinder" is a particular routing algorithm.  I think if you're solving an ILP, you're probably not doing the pathfinder algorithm?

Comment (stephenneuendorffer) [reply python/PythonPathFinderPass.cpp:35]: https://ieeexplore.ieee.org/document/1377269

**Pull request merged**
